packages/core/src/connectivity/scp-cf/destination/destination-from-service.ts

```diff
@@ -221,21 +227,69 @@ class DestinationFromServiceRetriever {
     return credentials[0];
   }
 
-  private async addOAuthSamlAuth(
-    destination: Destination,
-    destinationOrigin: DestinationOrigin
+  private async fetchDestinationByToken(
+    jwt: string | AuthAndExchangeTokens
   ): Promise<Destination> {
-    const destinationService = getService('destination');
+    const destinationService = getDestinationService();
 
-    if (!destinationService) {
+    return fetchDestination(
+      destinationService.credentials.uri,
+      jwt,
+      this.name,
+      this.options
+    );
+  }
+
+  private async addOAuth2UserTokenExchange(
+    destinationOrigin: DestinationOrigin
+  ): Promise<Destination> {
+    if (!this.options.userJwt) {
       throw Error(
-        `Failed to fetch destination "${this.name}"! No binding to a destination service found.`
+        'No user token (JWT) has been provided! This is strictly necessary for OAuth2UserTokenExchange.'
+      );
+    }
+    // This covers the three OAuth2UserTokenExchange cases https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/39d42654093e4f8db20398a06f7eab2b.html
+
+    // Case 1 Destination in provider and jwt issued for provider account
+    if (this.isProviderAndSubscriberSameTenant()) {
```

Comment:
Minor: I think this might be easier to read if there was a function that would get the `AuthAndExchangeTokens` before fetching the destination (you don't necessarily need an extra function for this).
Not that it is very hard to read, but there is boilerplate code in the way that could be avoided.

```ts
private async getAuthAndExchangeTokens(destinationOrigin: DestinationOrigin): Promise< AuthAndExchangeTokens> {
  // Option1, 2, 3
  // The magic happens here
}

private async addOAuth2UserTokenExchange(destinationOrigin: DestinationOrigin): Promise<Destination> {
  // ...
  return fetchDestination(getAuthAndExchangeTokens(destinationOrigin));
}
```

Reply:
good idea.